x/leverage/keeper/keeper.go

```diff
@@ -346,226 +346,83 @@ func (k Keeper) Decollateralize(ctx sdk.Context, borrowerAddr sdk.AccAddress, co
+		return sdk.Coin{}, sdk.Coin{}, sdk.Coin{}, err
 	}
 
-	if desiredReward.Amount.IsPositive() {
-		// user-controlled minimum ratio of reward to repayment, expressed in collateral base assets (not uTokens)
-		rewardTokenEquivalent, err := k.ExchangeUToken(ctx, reward)
-		if err != nil {
-			return sdk.ZeroInt(), sdk.ZeroInt(), err
-		}
-
-		minimumRewardRatio := sdk.NewDecFromInt(desiredReward.Amount).QuoInt(desiredRepayment.Amount)
-		actualRewardRatio := sdk.NewDecFromInt(rewardTokenEquivalent.Amount).QuoInt(repayment.Amount)
-		if actualRewardRatio.LT(minimumRewardRatio) {
-			return sdk.ZeroInt(), sdk.ZeroInt(), types.ErrLiquidationRewardRatio
-		}
-	}
-
-	// send repayment to leverage module account
-	if err = k.bankKeeper.SendCoinsFromAccountToModule(
-		ctx, liquidatorAddr,
-		types.ModuleName,
-		sdk.NewCoins(repayment),
-	); err != nil {
-		return sdk.ZeroInt(), sdk.ZeroInt(), err
+	// send base rewards from module to liquidator's account
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, liquidatorAddr, sdk.NewCoins(baseReward))
+	if err != nil {
+		return sdk.Coin{}, sdk.Coin{}, sdk.Coin{}, err
 	}
 
-	// update the remaining borrowed amount
-	owed := borrowed.AmountOf(repayment.Denom).Sub(repayment.Amount)
-	if err = k.setBorrow(ctx, borrowerAddr, sdk.NewCoin(repayment.Denom, owed)); err != nil {
-		return sdk.ZeroInt(), sdk.ZeroInt(), err
-	}
+	// get remaining collateral, ignoring blacklisted
+	remainingCollateral := k.filterAcceptedCoins(ctx, k.GetBorrowerCollateral(ctx, borrowerAddr))
 
-	// Reduce borrower collateral by reward amount
-	newBorrowerCollateral := sdk.NewCoin(reward.Denom, maxReward.Sub(reward.Amount))
-	if err = k.setCollateralAmount(ctx, borrowerAddr, newBorrowerCollateral); err != nil {
-		return sdk.ZeroInt(), sdk.ZeroInt(), err
-	}
-
-	// Send rewards to liquidator's account.
-	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, liquidatorAddr, sdk.NewCoins(reward))
-	if err != nil {
-		return sdk.ZeroInt(), sdk.ZeroInt(), err
-	}
-
-	// Detect bad debt (collateral == 0 after reward) for repayment by protocol reserves
-	if collateral.Sub(sdk.NewCoins(reward)).IsZero() {
-		for _, coin := range borrowed {
-			// Mark repayment denom as bad debt only if some debt remains after
-			// this liquidation. All other borrowed denoms were definitely not
-			// repaid in this liquidation so they are always marked as bad debt.
-			if coin.Denom != repayment.Denom || owed.IsPositive() {
-				if err := k.setBadDebtAddress(ctx, borrowerAddr, coin.Denom, true); err != nil {
-					return sdk.ZeroInt(), sdk.ZeroInt(), err
-				}
+	// detect bad debt if collateral is completely exhausted
+	if remainingCollateral.IsZero() {
+		for _, coin := range k.GetBorrowerBorrows(ctx, borrowerAddr) {
+			// set a bad debt flag for each borrowed denom
+			if err := k.setBadDebtAddress(ctx, borrowerAddr, coin.Denom, true); err != nil {
```

Comment:
Bad debt is currently just something to iterate over, so storing `0x01` is sufficient.

We could store the `amount` - but then it would be redundant with `adjusted borrow` which also stores amount. Adjusted borrow would also continue to accrue interest while this does not. Additionally, any operation (e.g. partial repay) that updated one of the two would have to keep the other in sync, while the bad debt existed.

Overall, it's simpler to avoid storing amount here.